From 2e4f4f9d110e293e30d83f5d641be6fbe6e2f118 Mon Sep 17 00:00:00 2001
From: Ananya Ghosh <ananghos@qti.qualcomm.com>
Date: Fri, 27 Feb 2026 15:10:30 +0530
Subject: [PATCH] arm_ffa: Add missing ffa_notifier_ops structure definition

The ffa_notifier_ops structure was referenced in struct ffa_ops but
not defined, causing compilation failures in drivers/tee/optee/ffa_abi.c
with errors:

  drivers/tee/optee/ffa_abi.c:821:43: error: invalid use of undefined
    type 'const struct ffa_notifier_ops'
  drivers/tee/optee/ffa_abi.c:839:48: error: invalid use of undefined
    type 'const struct ffa_notifier_ops'
  drivers/tee/optee/ffa_abi.c:863:43: error: invalid use of undefined
    type 'const struct ffa_notifier_ops'
  drivers/tee/optee/ffa_abi.c:982:26: error: invalid use of undefined
    type 'const struct ffa_notifier_ops'

Add the complete ffa_notifier_ops structure definition with all required
function pointers for notification handling:
- sched_recv_cb_register/unregister for scheduling callbacks
- notify_request/relinquish for notification management
- notify_send for sending notifications

Also add the required callback function typedefs (ffa_sched_recv_cb and
ffa_notifier_cb) that are used by the notifier operations.

This change is backported from upstream Linux kernel to fix the
compilation issue while maintaining compatibility with the existing
OP-TEE FFA ABI implementation.

Change-Id: Iaaa6d3016c6e46e876befd1b3dc859adcd95400e
Signed-off-by: Ananya Ghosh <ananghos@qti.qualcomm.com>
---
 include/linux/arm_ffa.h | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/include/linux/arm_ffa.h b/include/linux/arm_ffa.h
index cc060da51bec..144d30b59f96 100644
--- a/include/linux/arm_ffa.h
+++ b/include/linux/arm_ffa.h
@@ -367,10 +367,25 @@ struct ffa_mem_ops {
 	int (*memory_lend)(struct ffa_mem_ops_args *args);
 };
 
+typedef void (*ffa_sched_recv_cb)(u16 vcpu, bool is_per_vcpu, void *cb_data);
+typedef void (*ffa_notifier_cb)(int notify_id, void *cb_data);
+
+struct ffa_notifier_ops {
+	int (*sched_recv_cb_register)(struct ffa_device *dev,
+				      ffa_sched_recv_cb cb, void *cb_data);
+	int (*sched_recv_cb_unregister)(struct ffa_device *dev);
+	int (*notify_request)(struct ffa_device *dev, bool per_vcpu,
+			      ffa_notifier_cb cb, void *cb_data, int notify_id);
+	int (*notify_relinquish)(struct ffa_device *dev, int notify_id);
+	int (*notify_send)(struct ffa_device *dev, int notify_id, bool per_vcpu,
+			   u16 vcpu);
+};
+
 struct ffa_ops {
 	const struct ffa_info_ops *info_ops;
 	const struct ffa_msg_ops *msg_ops;
 	const struct ffa_mem_ops *mem_ops;
+	const struct ffa_notifier_ops *notifier_ops;
 };
 
 #endif /* _LINUX_ARM_FFA_H */
-- 
2.34.1

